File: Leetcode/Correct_A_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
 
    boolean flag = false;   
    public void calldfs (TreeNode curr, TreeNode parent, HashMap<TreeNode,Integer> dm, char dir)
    {
        if (flag)
            return;
        if (curr.left == null && curr.right == null)
            return;
        //System.out.println("current node is "+curr.val);
        int currHeight = dm.get(curr);
        if (curr.right != null)
        {
            int rightHeight = dm.get(curr.right);
            if (currHeight == rightHeight)
            {
                //System.out.println("this happened at node " + curr.val);
                if (dir =='L')
                    parent.left = null;
                else
                    parent.right = null;
                flag = true;
                return;
            }           
        }
        if (curr.left != null)
            calldfs (curr.left,curr,dm,'L');
        if (curr.right != null)
            calldfs (curr.right, curr, dm, 'R');
        return;
    }
    
    public void dfs (TreeNode curr, HashMap<TreeNode, Integer> dm, int level)
    {
        if (curr == null)
            return;
        dm.put (curr,level);
        dfs (curr.left, dm,level+1);
        dfs (curr.right, dm, level+1);
    }
    
    public TreeNode correctBinaryTree(TreeNode root) {
       flag = false;
       HashMap<TreeNode, Integer> dm = new HashMap();
       dfs (root, dm, 0);
       calldfs (root,root,dm,'L');
       return root; 
    }
}

File: Leetcode/Correct_A_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int repeated = -1;
    unordered_map<TreeNode*,int > umap;
    void dfs1 (TreeNode* curr, int level)
    {  
        if (flag || curr == nullptr)
            return;
        // cout<< "current node is " << curr->val << endl;
        if (umap[curr] !=0)
        {
           // cout << "inside here" << endl;
            umap[curr] = level;
            repeated = curr->val;
            flag = true;
            return;
        }
        umap[curr] = level;
        dfs1(curr->left,  level+1);
        dfs1(curr->right,  level+1);            
    }
    
    void dfs2(TreeNode* curr, TreeNode* parent, char ch)
    {
        if (curr == nullptr || flag)
            return;
        if (curr->right !=nullptr && curr->right->val == repeated && umap[curr] == umap[curr->right])
        {
            if (ch =='L')
                parent->left = NULL;
            else
                parent->right = NULL;
            flag = true;
            return;
        }
        dfs2(curr->left, curr, 'L');
        dfs2 (curr->right, curr,  'R');
        return;    
    }
    
    bool flag = false;    
    TreeNode* correctBinaryTree(TreeNode* root) {
        umap.clear();
        repeated = -1;
        flag = false;
        dfs1(root,1);
        flag = false;
        dfs2(root,root,'L');
        return root;
    }
};
